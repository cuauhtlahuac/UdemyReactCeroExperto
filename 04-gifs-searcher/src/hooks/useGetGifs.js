import { useState, useEffect } from 'react';

import { getGifs } from 'utils/getGifs';

export const useGetGifs = category => {
	const [ state, setState ] = useState({ data: [], loading: true });

	useEffect(
		() => {
			getGifs(category).then(imgs => {
				setState(
					{ data: imgs,
					  loading: false 
					});
			});
		},
		[ category ],
	);

	return state;
};
