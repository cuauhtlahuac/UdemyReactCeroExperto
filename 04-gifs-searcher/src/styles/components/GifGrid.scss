.image-grid {
	align-items: baseline;
	display: grid;
	grid-gap: 10px;
    grid-template-columns: repeat(auto-fit, minmax(16em, 1fr));
    row-gap: 15px;
	min-width: 0;
}

.loading {
	font-size: 2rem;
	font-family: 'Rubik' Arial, Helvetica, sans-serif;
	font-weight: 500;
	color: rgba(0, 0, 0, 0.274);
}