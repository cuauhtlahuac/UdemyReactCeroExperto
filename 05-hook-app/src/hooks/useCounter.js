import { useState } from "react"

export default function useCounter( initialState = 13 ) {
    const [counter, setCounter] = useState( initialState )

    function decrement(factor = 1) {
        setCounter(( counter <= 0 ) ? 0 : counter - factor);
    };

    function reset() {
        setCounter( initialState );
    };

    function increment(factor = 1) {
        setCounter(counter + factor);
    };

    return {
        counter,
        decrement,
        reset,
        increment,
    }
}
