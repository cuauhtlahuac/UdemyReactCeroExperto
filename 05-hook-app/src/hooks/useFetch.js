import { useState, useEffect, useRef } from 'react';

export default function useFetch( url ) {

	const initialState = {
		data: null,
		loading: true,
		error: null,
	};

	const [ state, setState ] = useState( initialState );

	const isMounted = useRef(true);

	useEffect(() => {
		
		return () => {
			isMounted.current = false;
		}
	}, [isMounted])

	useEffect(() => {
		setState({
			data: null,
			loading: true,
			error: null,
		});

        fetch( url )
        .then( resp => resp.json() )
        .then( data => { 

			if(isMounted.current){
				setState({
					data,
					loading: false,
					error: null,
				})
			};
		}).catch( error => (setState({
			data: null,
			loading: false,
			error: error,
		})))
	}, [ url ]);

	return state;
}
