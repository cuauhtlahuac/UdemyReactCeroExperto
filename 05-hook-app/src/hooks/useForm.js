import { useState, useEffect } from 'react';

export const useForm = (initialState = {}) => {
	const [ values, setValues ] = useState(initialState);
	const [ current, setCurrent ] = useState();

	useEffect(
		() => {

			if (current) {
				document.querySelector(
					`#simpleFormWithHooks [for=${current}]`,
				).style.color = 'limegreen';
			}
		},
		[ current ],
	);

	const handleInputChange = ({ target }) => {
		setValues({
			...values,
			[target.name]: target.value,
		});
		setCurrent(target.name);
	};

	return [ values, handleInputChange ];
};
