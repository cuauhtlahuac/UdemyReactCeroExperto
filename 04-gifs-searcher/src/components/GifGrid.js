import React, { lazy, Suspense } from 'react';
import PropTypes from 'prop-types';

import { useGetGifs } from 'hooks/useGetGifs';

const GifGridItem = lazy(() => import('./GifGridItem'));

const Loader = () => <div className="loading">Loading...</div>;

const GifGrid = ({ category }) => {
	const { data: images, loading } = useGetGifs(category);

	return (
		<section>
			<h3 className=
			"title animate__animated
			 animate__slideInRight"
			 >
				{category}
			</h3>

			{loading && (
				<p className=
				"loading animate__animated
				animate__fadeIn animate__repeat-3
				animate__faster"
				>
					loading...
				</p>
			)}

			<section className="image-grid">
				{images.map(image => (
					<Suspense fallback={ <Loader /> } key={image.id}>
						<GifGridItem {...image} />
					</Suspense>
				))}
			</section>
		</section>
	);
};

export default GifGrid;

GifGrid.propTypes = {
	category: PropTypes.string,
};
