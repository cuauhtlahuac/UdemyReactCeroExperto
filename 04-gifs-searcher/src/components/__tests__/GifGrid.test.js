import React from 'react'
import { shallow } from "enzyme"

import GifGrid from 'components/GifGrid'
import { useGetGifs } from 'hooks/useGetGifs'

jest.mock('hooks/useGetGifs'); // con esto se finge cualquier llamada a la funcion interna

describe('GifGrid tests', () => {

    test('should render correctly', () => {

        useGetGifs.mockReturnValue({
            data:[],
            load: true,
        }); // simula el return de la funcion

        const component = shallow( <GifGrid category="sol"/> );
        expect( component ).toMatchSnapshot();
    });

    test('should render with mock change', () => {
        const gifs = [
            {
                id: 'ass156',
                title: "title",
                link: "http://any.com.any",
                gif: "http://any.com.gif",
                webp_size: 5695,        
            }];

        useGetGifs.mockReturnValue({
            data: gifs,
            load: false,
        }); 

        const component = shallow( <GifGrid category="luna"/> );
        expect( component ).toMatchSnapshot();

    })
    
    
    
})
