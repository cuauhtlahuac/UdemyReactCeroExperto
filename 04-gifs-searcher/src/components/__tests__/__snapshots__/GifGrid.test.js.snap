// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GifGrid tests should render correctly 1`] = `
<section>
  <h3
    className="title animate__animated animate__slideInRight"
  >
    sol
  </h3>
  <section
    className="image-grid"
  />
</section>
`;

exports[`GifGrid tests should render with mock change 1`] = `
<section>
  <h3
    className="title animate__animated animate__slideInRight"
  >
    luna
  </h3>
  <section
    className="image-grid"
  >
    <Suspense
      fallback={<Loader />}
      key="ass156"
    >
      <lazy
        gif="http://any.com.gif"
        id="ass156"
        link="http://any.com.any"
        title="title"
        webp_size={5695}
      />
    </Suspense>
  </section>
</section>
`;
